kart/completion.py

```diff
@@ -0,0 +1,153 @@
+    script = _completion_scripts.get(shell)
+    if script is None:
+        click.echo(f"Shell {shell} not supported.", err=True)
+        sys.exit(1)
+    return (
+        script
+        % dict(
+            complete_func="_{}_completion".format(cf_name),
+            complete_var=complete_var,
+            prog_name=prog_name,
+        )
+    ).strip()
+
+
+def install_bash(*, prog_name: str, complete_var: str, shell: str) -> Path:
+    # Ref: https://github.com/scop/bash-completion#faq
+    # It seems bash-completion is the official completion system for bash:
+    # Ref: https://www.gnu.org/software/bash/manual/html_node/A-Programmable-Completion-Example.html
+    # But installing in the locations from the docs doesn't seem to have effect
+    completion_path = Path.home() / f".bash_completions/{prog_name}.sh"
+    rc_path = Path.home() / ".bashrc"
+    rc_path.parent.mkdir(parents=True, exist_ok=True)
+    rc_content = ""
+    if rc_path.is_file():
+        rc_content = rc_path.read_text()
+    completion_init_lines = [f"source {completion_path}"]
+    for line in completion_init_lines:
+        if line not in rc_content:
+            rc_content += f"\n{line}"
+    rc_content += "\n"
+    rc_path.write_text(rc_content)
+    # Install completion
+    completion_path.parent.mkdir(parents=True, exist_ok=True)
+    script_content = get_completion_script(
+        prog_name=prog_name, complete_var=complete_var, shell=shell
+    )
+    completion_path.write_text(script_content)
+    return completion_path
+
+
+def install_zsh(*, prog_name: str, complete_var: str, shell: str) -> Path:
+    # Setup Zsh and load ~/.zfunc
+    zshrc_path = Path.home() / ".zshrc"
+    zshrc_path.parent.mkdir(parents=True, exist_ok=True)
+    zshrc_content = ""
+    if zshrc_path.is_file():
+        zshrc_content = zshrc_path.read_text()
+    completion_init_lines = [
+        "autoload -Uz compinit",
+        "zstyle ':completion:*' menu select",
+        "fpath+=~/.zfunc; compinit",
+    ]
+    for line in completion_init_lines:
+        if line not in zshrc_content:
+            zshrc_content += f"\n{line}"
+    zshrc_content += "\n"
+    zshrc_path.write_text(zshrc_content)
+    # Install completion under ~/.zfunc/
+    path_obj = Path.home() / f".zfunc/_{prog_name}"
+    path_obj.parent.mkdir(parents=True, exist_ok=True)
+    script_content = get_completion_script(
+        prog_name=prog_name, complete_var=complete_var, shell=shell
+    )
+    path_obj.write_text(script_content)
+    return path_obj
+
+
+def install_fish(*, prog_name: str, complete_var: str, shell: str) -> Path:
+    path_obj = Path.home() / f".config/fish/completions/{prog_name}.fish"
+    parent_dir: Path = path_obj.parent
+    parent_dir.mkdir(parents=True, exist_ok=True)
+    script_content = get_completion_script(
+        prog_name=prog_name, complete_var=complete_var, shell=shell
+    )
+    path_obj.write_text(f"{script_content}\n")
+    return path_obj
+
+
+def install(
```

Comment:
I see there's an install method in click_completion that maybe does something like this and that you don't seem to be calling - but it looks like maybe it's not as good as your one, ie, it doesn't check if it's already installed. Can you confirm?

Reply:
Yes, there's code in [`click-contrib/click-completion`](https://github.com/click-contrib/click-completion/) which does something similar to this (ref [code](https://github.com/click-contrib/click-completion/blob/a9b705d4f4df4579a96851b2b14b0c59765ef6ce/click_completion/core.py#L319-L381)). I have decided to not use [`click-contrib/click-completion`](https://github.com/click-contrib/click-completion/) because it seems to predate [Click's shell compeletion](https://click.palletsprojects.com/en/8.1.x/shell-completion/) capabilites (see [click-contrib/click-completion#41 (issue)](https://github.com/click-contrib/click-completion/issues/41)).